Enabled Detekt with Baseline files

Armaxis opened the pull request:
Added Detekt with a default configuration and created Baseline files for all subprojects

Currently uses standard ruleset, except for the rule about manually calling garbage collection

Removed Checkstyle

Resolves #1396

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/square/leakcanary?pullRequest=1535) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/square/leakcanary?pullRequest=1535) before we can accept your contribution.<br/><hr/>**Artem Chubaryan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/square/leakcanary?pullRequest=1535) it.</sub>

pyricau commented:
The top commit (c0551443e37c101abe7044327cc2dc5750f8808b) is not recognized by GitHub as authored by you. You probably used a different email address to author the commit. You could amend the commit and force push, or you can add that email address to your known addresses in GitHub.

pyricau commented:
Thanks!

How do I generate / override the baseline? Btw I hadn't realized we'd have so many things in the baseline, that's a lot a content committed :)

Armaxis commented:
- I updated the email, now CLA is ok! :) 
- Travis Build still fails. I heard that there was an update from Google who pushed new version of sdk 29 which broke a lot of travis builds, so I rebuild the branch, but it still fails :(
- To generate / update detekt baseline run `gradlew detektBaseline` to refresh for all the files or run task inside a particular module to refresh only one. 
- Right now in `detekt-config.yml` there's a setting that says `maxIssues: 10` - so if these are few issues build will be considered successful. Might want to consider lowering it to 1 so nothing goes unnoticed.
- Yes, there's a lot of violations right now; but if we disable some of the rules there will be a lot less of those :)

pyricau commented:
Do you have any suggestions on rules we should disable?

Armaxis commented on leakcanary-object-watcher-android/src/main/java/leakcanary/internal/DefaultCanaryLog.kt:
```diff
@@ -9,17 +9,12 @@ internal class DefaultCanaryLog : Logger {
     if (formatted.length < 4000) {
       Log.d("LeakCanary", formatted)
     } else {
-      val lines = formatted.split("\n".toRegex())
```
Regular expression was getting compiled every time, so I extracted it to companion object. Also removed redundant conversion to typed array

pyricau replied:
wow, nice catch!

Armaxis commented on leakcanary-object-watcher/src/main/java/leakcanary/ObjectWatcher.kt:
```diff
@@ -134,11 +134,10 @@ class ObjectWatcher constructor(
-        ((if (watchedObject is Class<*>) watchedObject.toString() else "instance of ${watchedObject.javaClass.name}") + if (name.isNotEmpty()) " named $name" else ""),
-        key
-    )
+    val nameForLogging =
```
Not sure if it's the best name for this variable

Armaxis commented:
Following checks can be disabled:

- **NewLineAtEndOfFile** - **disabled** (it's a matter of taste, not important)
- **MagicNumber** - **disabled** (too many of those, and the codebase's nature calls for the usage of lots of magic numbers, so it's fine)
- **ExplicitGarbageCollectionCall** - **disabled** (yeah, we want to do that :))
- **TooGenericExceptionCaught** - **disabled** (again, for this library it makes sense)
- **TooGenericExceptionThrown** - **disabled** (same)

For these checks, I adjusted thresholds so they are more relaxed but still capture extreme violations. Some might be relaxed even further.

- **ReturnCount** - increased from 2 to 4
- **LongMethod** - increased from 60 to 90; ignored default params
- **ComplexMethod** - increased from 10 to 30; ignored 'when' entries
- **NestedBlockDepth** - increased from 4 to 7
- **TooManyFunctions** - increased from 11 to 12
- **MaxLineLength** - increased from 120 to 150
- **EmptyFunctionBlock** - allowed empty overridden functions

With these changes to rules, most of the issues were addressed. I went ahead and also fixed some of the remaining things, like usage of **spread operator**, **safe cast** and **max line**. 
The remaining few violations were **added to the baseline** - those are reasonable considering the specifics of the implementation. Pushed updated configuration and baseline

pyricau commented:
Thanks so much!

pyricau commented on leakcanary-android-sample/detekt-baseline.xml:
```diff
@@ -0,0 +1,5 @@
+<SmellBaseline>
+  <Blacklist></Blacklist>
+  <Whitelist></Whitelist>
+</SmellBaseline>
```
do we need empty files?

Armaxis replied:
I believe we do, so it can update those later if we add more exclusions.
Will double-check it.

Armaxis replied:
I've got a better plan! 
Instead of adding baseline files I would just annotate the existing issues with `@Suppress`. This makes more sense since they aren't the issues we will be fixed in the future; they are completely ok the way they are.

This way there will be no baseline files at all. What do you think?

pyricau commented on leakcanary-object-watcher/src/main/java/leakcanary/ObjectWatcher.kt:
```diff
@@ -134,11 +134,10 @@ class ObjectWatcher constructor(
+    val nameForLogging =
+      (if (watchedObject is Class<*>) watchedObject.toString() else "instance of ${watchedObject.javaClass.name}") +
+      (if (name.isNotEmpty()) " named $name" else "")
+    SharkLog.d("Watching $nameForLogging with key $key")
```
The previous log was bad, this makes it slightly worst :) . Context: the idea with these logs is that they're no-op when there's no logger, and there's no unnecessary string concatenation. Here the log was actually bad, it should have done something like `if (SharkLog.logger != null) {}` (or maybe SharkLog should have a util method) to avoid any object construction. `"Watching $nameForLogging with key $key"` involves string interpolation even when there is no logger.

This is not a major impact, but more of a good practice.

Armaxis replied:
Ooh, I see! I'll restructure in a better way later today to prevent this unnecessary allocation.

On a side note: the previous call to `SharkLog.d` was passing the arguments that were later passed into the `String.format()` to replace `%s` placeholders. That internally uses `Matchers` and `Formatter` which most likely have worse performance than using Kotlin's string template (which uses `StringBuilder`)
So I wonder if it would be better to replace the calls like `logger.d("Logging %s", value)` with `logger.d("Logging $value")`

pyricau replied:
It would be more efficient when logger is set, but less efficient when logger is not set.

Maybe instead SharkLog's API should be changed to take an inline lambda that returns the formatted string:

```
SharkLog.d {
"Logging $value"
}
```

inline means no extra cost, you get the kotlin string templates, and if there's no logger then there's no cost.

Armaxis replied:
Wouldn't it be more expensive to create an instance of that lambda?

```
//SharkLog.kt:
fun d(lambda:() -> String) {
  val logger = logger ?: return
  logger.d(lambda.invoke())
}

//Test class:
private fun getI() = 6

@Test fun logTest() {
  val i = getI()
  SharkLog.d { "oops$i" }
  SharkLog.d ("oops$i")
}

//Decompiled Java
@Test
public final void logTest() {
   final int i = this.getI();
   SharkLog.INSTANCE.d((Function0)(new Function0() {
      // $FF: synthetic method
      // $FF: bridge method
      public Object invoke() {
         return this.invoke();
      }
      @NotNull
      public final String invoke() {
         return "oops" + i;
      }
   }));
   SharkLog.INSTANCE.d("oops" + i, new Object[0]);
}
private final int getI() {
   return 6;
}
```
Or am I doing something wrong here?

pyricau replied:
Make `d` an inline function and then the lambda gets inlined

Armaxis replied:
Aha! Way better now! 
I keep forgetting about this little magic word 😄 

```
@Test
public final void logTest() {
   int i = this.getI();
   SharkLog this_$iv = SharkLog.INSTANCE;
   int $i$f$d = false;
   Logger var10000 = this_$iv.getLogger();
   if (var10000 != null) {
      Logger logger$iv = var10000;
      int var5 = false;
      String var7 = "oops" + i;
      logger$iv.d(var7, new Object[0]);
   }
   SharkLog.INSTANCE.d("oops" + i, new Object[0]);
}
```

pyricau replied:
with that we can get rid of the vararg and lose the `new Object[0]`

Armaxis replied:
I will make it as a separate Pull Request, probably tomorrow

pyricau commented on shark-cli/src/main/java/shark/Main.kt:
```diff
@@ -124,7 +124,7 @@ private fun runCommand(
   directory: File,
   vararg arguments: String
 ): String {
-  val process = ProcessBuilder(*arguments)
```
what's wrong with the spread operator btw?

Armaxis replied:
From the [rule description:](https://arturbosch.github.io/detekt/performance.html#spreadoperator)

> SpreadOperator
In most cases using a spread operator causes a full copy of the array to be created before calling a method. This has a very high performance penalty. Benchmarks showing this performance penalty can be seen here: https://sites.google.com/a/athaydes.com/renato-athaydes/posts/kotlinshiddencosts-benchmarks
> The Kotlin compiler since v1.1.60 has an optimization that skips the array copy when an array constructor function is used to create the arguments that are passed to the vararg parameter. When type resolution is enabled in detekt this case will not be flagged by the rule since it doesn’t suffer the performance penalty of an array copy.

In this particular case, it doesn't really matter. Using spread operator will result in invoking this constructor, that loops over arguments and adds commands one by one:

```
public ProcessBuilder(String... command) {
    this.command = new ArrayList<>(command.length);
    for (String arg : command)
        this.command.add(arg);
}
```
While converting them to lo list via `.asList()` will first use `ArraysUtilJVM.asList()` to convert array to list and then it will call constructor which will use the list:
```
public ProcessBuilder(List<String> command) {
    if (command == null)
        throw new NullPointerException();
    this.command = command;
}
```

pyricau commented on leakcanary-object-watcher-android/src/main/java/leakcanary/internal/DefaultCanaryLog.kt:
```diff
@@ -30,10 +25,10 @@ internal class DefaultCanaryLog : Logger {
-            throwable
-        )
-    )
+    d("${String.format(message, args)}\n${Log.getStackTraceString(throwable)}")
```
Are you 100% that we don't need the spread operator here? String.format takes an object, and args is an object, so I'm worried it's going to run toString on the args array itself, hence the spread operator

pyricau replied:
Should be the same method yes. I've definitely had this kind of issues in java in the past so I'm worried :)

Armaxis replied:
You were absolutely right! Even though it navigates to the same method it doesn't behave in the same way and results in an exception:
```
@Test fun loggerTest() {
  d("Too many placeholders %s %s %s", "one", "two", "three")
}
fun d(message: String, vararg args: Any?) {
  val formatted = String.format(message, args)
  println(formatted)
}

java.util.MissingFormatArgumentException: Format specifier '%s'
	at java.util.Formatter.format(Formatter.java:2519)
	at java.util.Formatter.format(Formatter.java:2455)
	at java.lang.String.format(String.java:2940)
```

I looked into the decompiled code, and here's what it's doing (omitted unrelated code):
```
public final void d(@NotNull String message, @NotNull Object... args)

//when using String.format(message, args)
Object[] var5 = new Object[]{args};
String formatted = String.format(message, Arrays.copyOf(var5, var5.length));
System.out.println(formatted);

//when using String.format(message, *args)
Object[] var5 = Arrays.copyOf(args, args.length);
String formatted = String.format(message, Arrays.copyOf(var5, var5.length));
System.out.println(formatted);
```
`Object[] var5 = new Object[]{args};` this resulted in creating another array with array inside of it:

![image](https://user-images.githubusercontent.com/2152942/63559718-da55a680-c507-11e9-8e88-90aa43cd824b.png)
****

**This raises two questions:**

1. Should I just revert things related to spread operator back and disable the check? It might cause more hidden problems like this one. Furthermore, if `SharkLog` gets a lambda parameter, the spread operator will not be needed.
2. Is it a bug in the Kotlin plugin or somewhere that it points to the same method but ends up to be different things? I feel like it should be raised as an issue, what do you think?

pyricau replied:
1. => How about we revert / disable the check for now, and then follow up with the inline sharklog?
2. Probably worth filing. There might be some cases where this makes sense, but at the very least there should be more education on this.

Armaxis replied:
1. Will do.

2. https://discuss.kotlinlang.org/t/report-missing-spread-operator-for-vararg/14003

Armaxis commented:
Couple more pending items:
- Replace **Baseline** files with `@Suppress` annotations?
- Do we need the statistics report per module? It doesn't seem to be affecting the performance of report, but it makes logs bigger adding following:
>Complexity Report:
	- 2130 lines of code (loc)
	- 1511 source lines of code (sloc)
	- 1084 logical lines of code (lloc)
	- 373 comment lines of code (cloc)
	- 317 McCabe complexity (mcc)
	- 0 number of total code smells
	- 24 % comment source ratio
	- 292 mcc per 1000 lloc
	- 0 code smells per 1000 lloc
Project Statistics:
	- number of properties: 221
	- number of functions: 61
	- number of classes: 64
	- number of packages: 2
	- number of kt files: 19
- The default configuration [allows having up to 10 weighted issues](https://arturbosch.github.io/detekt/failonbuild.html) (per module) for a build to succeed. On my other project I set it to be **1** - so that build fails whenever there's an issue.
But there's a backside to this: When Travis is running `detekt` on the root level project it executes `detekt` sequentially module by module. If one of the modules fails - it stops and never runs it on the next module. 
The workaround for it is running `detekt` locally and fixing all the issues before pushing to remote. There's an ongoing discussion with some support for multi-project available, but it's non-trivial setup: https://github.com/arturbosch/detekt/issues/1295 
So, after all, we can point this out in contribution guidelines, something like:
Run `./gradlew check` before pushing the code and resolve (or suppress) all the reported issues.

pyricau commented:
I like the stats reports per module! Let's keep it.

Maybe we can make `./gradlew check` a commit hook?

Armaxis commented:
> I like the stats reports per module! Let's keep it.

Ok!

> Maybe we can make `./gradlew check` a commit hook?

In my other project, I made it as a push hook - this way it doesn't slow down local development and usually the issues are pretty trivial, so fixing them before pushing doesn't take long. 
Although having a commit hook has its own benefits, like finding issues earlier.

pyricau commented:
oh sorry yeah I actually should have said push hook :)

Armaxis commented:
> oh sorry yeah I actually should have said push hook :)

Can I make this as a separate pull request? Last time I did this hook I had to do some magic to make it work both on macs and windows, I don't want to jeopardize this PR with that stuff :)

Armaxis commented:
Updated Pull request with all the changes
